packages-oss.yml

```diff
@@ -0,0 +1,282 @@
+layers:
+  - name: base
+    dockerfile: |-
+      RUN apt-get update -y && apt-get install --no-install-recommends -y -q \
+                               curl \
+                               zip \
+                               build-essential \
+                               gcc-multilib \
+                               g++-multilib \
+                               ca-certificates \
+                               git mercurial bzr \
+                               gnupg \
+                               libltdl-dev \
+                               libltdl7 \
+      						             bash \
+                            && rm -rf /var/lib/apt/lists/*
+  - name: install-go
+    dockerfile: |-
+      ENV GOPATH /gopath
+      ENV GOROOT /goroot
+      RUN mkdir $GOROOT && mkdir $GOPATH
+      RUN curl https://storage.googleapis.com/golang/go{{.GO_VERSION}}.linux-amd64.tar.gz \
+                 | tar xzf - -C $GOROOT --strip-components=1
+      ENV PATH $GOROOT/bin:$GOPATH/bin:$PATH
+  - name: install-go-tools
+    dockerfile: |
+      ENV GO111MODULE=off
+      RUN go get golang.org/x/tools/cmd/goimports
+      RUN go get github.com/hashicorp/go-bindata
+      RUN go get github.com/hashicorp/go-bindata/go-bindata
+      RUN go get github.com/elazarl/go-bindata-assetfs
+      RUN go get github.com/elazarl/go-bindata-assetfs/go-bindata-assetfs
+  - name: set-workdir
+    dockerfile: |
+      ENV REPO=github.com/hashicorp/vault
+      ENV DIR=$GOPATH/src/$REPO
+      RUN mkdir -p $DIR
+      WORKDIR $DIR
+  - name: install-yarn
+    dockerfile: |-
+      RUN curl -sL https://deb.nodesource.com/setup_10.x | bash -
+      RUN curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+      RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+      RUN apt-get update -y && apt-get install -y -q nodejs yarn={{.YARN_VERSION}} \
+                            && rm -rf /var/lib/apt/lists/*
+  - name: ui-dependencies
+    source-include: ui/package.json ui/yarn.lock
+    dockerfile: |-
+      RUN cd ui && yarn install
+      RUN cd ui && npm rebuild node-sass
+  - name: build-ui
+    source-include: ui/
+    dockerfile: |-
+      RUN cd ui && yarn run build
+  - name: build-static-assets
+    source-include: Makefile
+    dockerfile: |-
+      RUN go-bindata-assetfs -o bindata_assetfs.go -pkg http -prefix pkg -modtime 1480000000 -tags ui ./pkg/web_ui/...
+      RUN mkdir -p http && mv bindata_assetfs.go http/
+      RUN goimports -w http/bindata_assetfs.go
+  # Up to now, the layers do not vary between packages. This means the layers above
+  # can be built prior to package builds in order to warm the cache and speed up
+  # builds.
+  #
+  # warm-go-build-vendor-cache does vary based on GOOS, GOARCH and CGO_ENABLED though.
+  # This means we should cache this in CI per package build.
+  - name: warm-go-build-vendor-cache
```

Comment:
What's the purpose of building the vendored packages?

Reply:
This allows us to populate the build cache with the results of building the vendor packages. Whenever a new build happens that does not touch the vendor directory, the build is greatly speeded up as it only needs to rebuild our code.

Reply:
Couldn't we do that for the build as a whole, rather than targeting just vendor/ ?

Reply:
In usual circumstances that would work, however the approach packagespec takes is to segregate the source code into discrete chunks. In this case, the layer being built does not contain the application code at all, it just contains the vendor directory. This is done to improve efficiency. Because we know that only the vendor directory is present, we can use the Git commit SHA of the last commit to have touched the `vendor/` tree as part of the cache key. Now, future builds that don't touch `vendor/` can be run in the docker image that already built the code in vendor.

Reply:
The structure of the cache reflects the structure of the repository, as defined by the git pathspec `source-include` definitions in each layer. Therefore different branches all just get the right version of the cache for each layer according to the git commit that touched that part of the tree last. (The entire contents of the rendered layer Dockerfile as well as the `source-include/source-exclude` is also taken into account in cache keys, of course.)

Upcoming plans for more intensive testing of our builds / performing release builds much more frequently than we do now should benefit significantly from this more granular caching.